# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webaudio/FFTBlock.h
# Commit: 9973c75413d6
# Full Hash: 9973c75413d6237b9005b9ab908898a153bd2f06
# Author: Paul Adenot <paul@paul.cx>
# Date: 2024-02-28 04:30:28
# Regressor Bug: 1879873
# File Overlap Count: 1
# Description:
#   Bug 1879873 - Remove MOZ_LIBAV_FFT. r=karlt
#   
#   We have some time to use av_fft.h, after which we'll have to use tx.h directly.
#   It's not hard to update.
#   
# ==============================================================================

diff -r 7c947114da80 -r 9973c75413d6 dom/media/webaudio/FFTBlock.h
--- a/dom/media/webaudio/FFTBlock.h	Tue Feb 27 12:40:22 2024 +0000
+++ b/dom/media/webaudio/FFTBlock.h	Tue Feb 27 12:40:22 2024 +0000
@@ -7,31 +7,17 @@
 #ifndef FFTBlock_h_
 #define FFTBlock_h_
 
-#ifdef BUILD_ARM_NEON
-#  include <cmath>
-#  include "mozilla/arm.h"
-#  include "dl/sp/api/omxSP.h"
-#endif
-
 #include "AlignedTArray.h"
 #include "AudioNodeEngine.h"
-#if defined(MOZ_LIBAV_FFT)
-#  include "FFmpegRDFTTypes.h"
-#  include "FFVPXRuntimeLinker.h"
-#else
-#  include "kiss_fft/kiss_fftr.h"
-#endif
+#include "FFmpegRDFTTypes.h"
+#include "FFVPXRuntimeLinker.h"
 
 namespace mozilla {
 
 // This class defines an FFT block, loosely modeled after Blink's FFTFrame
 // class to make sharing code with Blink easy.
-// Currently it's implemented on top of KissFFT on all platforms.
 class FFTBlock final {
   union ComplexU {
-#if !defined(MOZ_LIBAV_FFT)
-    kiss_fft_cpx c;
-#endif
     float f[2];
     struct {
       float r;
@@ -41,28 +27,13 @@
 
  public:
   static void MainThreadInit() {
-#ifdef MOZ_LIBAV_FFT
     FFVPXRuntimeLinker::Init();
     if (!sRDFTFuncs.init) {
       FFVPXRuntimeLinker::GetRDFTFuncs(&sRDFTFuncs);
     }
-#endif
   }
 
-  explicit FFTBlock(uint32_t aFFTSize)
-#if defined(MOZ_LIBAV_FFT)
-      : mAvRDFT(nullptr),
-        mAvIRDFT(nullptr)
-#else
-      : mKissFFT(nullptr),
-        mKissIFFT(nullptr)
-#  ifdef BUILD_ARM_NEON
-        ,
-        mOmxFFT(nullptr),
-        mOmxIFFT(nullptr)
-#  endif
-#endif
-  {
+  explicit FFTBlock(uint32_t aFFTSize) : mAvRDFT(nullptr), mAvIRDFT(nullptr) {
     MOZ_COUNT_CTOR(FFTBlock);
     SetFFTSize(aFFTSize);
   }
@@ -83,22 +54,11 @@
       return;
     }
 
-#if defined(MOZ_LIBAV_FFT)
     PodCopy(mOutputBuffer.Elements()->f, aData, mFFTSize);
     sRDFTFuncs.calc(mAvRDFT, mOutputBuffer.Elements()->f);
     // Recover packed Nyquist.
     mOutputBuffer[mFFTSize / 2].r = mOutputBuffer[0].i;
     mOutputBuffer[0].i = 0.0f;
-#else
-#  ifdef BUILD_ARM_NEON
-    if (mozilla::supports_neon()) {
-      omxSP_FFTFwd_RToCCS_F32_Sfs(aData, mOutputBuffer.Elements()->f, mOmxFFT);
-    } else
-#  endif
-    {
-      kiss_fftr(mKissFFT, aData, &(mOutputBuffer.Elements()->c));
-    }
-#endif
   }
   // Inverse-transform internal data and store the resulting FFTSize()
   // points in aDataOut.
@@ -116,27 +76,13 @@
       return;
     };
 
-#if defined(MOZ_LIBAV_FFT)
-    {
-      // Even though this function doesn't scale, the libav forward transform
-      // gives a value that needs scaling by 2 in order for things to turn out
-      // similar to how we expect from kissfft/openmax.
-      AudioBufferCopyWithScale(mOutputBuffer.Elements()->f, 2.0f, aDataOut,
-                               mFFTSize);
-      aDataOut[1] = 2.0f * mOutputBuffer[mFFTSize / 2].r;  // Packed Nyquist
-      sRDFTFuncs.calc(mAvIRDFT, aDataOut);
-    }
-#else
-#  ifdef BUILD_ARM_NEON
-    if (mozilla::supports_neon()) {
-      omxSP_FFTInv_CCSToR_F32_Sfs_unscaled(mOutputBuffer.Elements()->f,
-                                           aDataOut, mOmxIFFT);
-    } else
-#  endif
-    {
-      kiss_fftri(mKissIFFT, &(mOutputBuffer.Elements()->c), aDataOut);
-    }
-#endif
+    // Even though this function doesn't scale, the libav forward transform
+    // gives a value that needs scaling by 2 in order for things to turn out
+    // similar to how we expect from kissfft/openmax.
+    AudioBufferCopyWithScale(mOutputBuffer.Elements()->f, 2.0f, aDataOut,
+                             mFFTSize);
+    aDataOut[1] = 2.0f * mOutputBuffer[mFFTSize / 2].r;  // Packed Nyquist
+    sRDFTFuncs.calc(mAvIRDFT, aDataOut);
   }
 
   void Multiply(const FFTBlock& aFrame) {
@@ -188,7 +134,6 @@
   size_t SizeOfExcludingThis(MallocSizeOf aMallocSizeOf) const {
     size_t amount = 0;
 
-#if defined(MOZ_LIBAV_FFT)
     auto ComputedSizeOfContextIfSet = [this](void* aContext) -> size_t {
       if (!aContext) {
         return 0;
@@ -209,17 +154,6 @@
 
     amount += ComputedSizeOfContextIfSet(mAvRDFT);
     amount += ComputedSizeOfContextIfSet(mAvIRDFT);
-#else
-#  ifdef BUILD_ARM_NEON
-    amount += aMallocSizeOf(mOmxFFT);
-    amount += aMallocSizeOf(mOmxIFFT);
-#  endif
-#  ifdef USE_SIMD
-#    error kiss fft uses malloc only when USE_SIMD is not defined
-#  endif
-    amount += aMallocSizeOf(mKissFFT);
-    amount += aMallocSizeOf(mKissIFFT);
-#endif
     amount += mOutputBuffer.ShallowSizeOfExcludingThis(aMallocSizeOf);
     return amount;
   }
@@ -233,7 +167,6 @@
   void operator=(const FFTBlock& other) = delete;
 
   bool EnsureFFT() {
-#if defined(MOZ_LIBAV_FFT)
     if (!mAvRDFT) {
       if (!sRDFTFuncs.init) {
         return false;
@@ -241,25 +174,10 @@
 
       mAvRDFT = sRDFTFuncs.init(FloorLog2(mFFTSize), DFT_R2C);
     }
-#else
-#  ifdef BUILD_ARM_NEON
-    if (mozilla::supports_neon()) {
-      if (!mOmxFFT) {
-        mOmxFFT = createOmxFFT(mFFTSize);
-      }
-    } else
-#  endif
-    {
-      if (!mKissFFT) {
-        mKissFFT = kiss_fftr_alloc(mFFTSize, 0, nullptr, nullptr);
-      }
-    }
-#endif
     return true;
   }
 
   bool EnsureIFFT() {
-#if defined(MOZ_LIBAV_FFT)
     if (!mAvIRDFT) {
       if (!sRDFTFuncs.init) {
         return false;
@@ -267,44 +185,10 @@
 
       mAvIRDFT = sRDFTFuncs.init(FloorLog2(mFFTSize), IDFT_C2R);
     }
-#else
-#  ifdef BUILD_ARM_NEON
-    if (mozilla::supports_neon()) {
-      if (!mOmxIFFT) {
-        mOmxIFFT = createOmxFFT(mFFTSize);
-      }
-    } else
-#  endif
-    {
-      if (!mKissIFFT) {
-        mKissIFFT = kiss_fftr_alloc(mFFTSize, 1, nullptr, nullptr);
-      }
-    }
-#endif
     return true;
   }
 
-#ifdef BUILD_ARM_NEON
-  static OMXFFTSpec_R_F32* createOmxFFT(uint32_t aFFTSize) {
-    MOZ_ASSERT((aFFTSize & (aFFTSize - 1)) == 0);
-    OMX_INT bufSize;
-    OMX_INT order = FloorLog2(aFFTSize);
-    MOZ_ASSERT(aFFTSize >> order == 1);
-    OMXResult status = omxSP_FFTGetBufSize_R_F32(order, &bufSize);
-    if (status == OMX_Sts_NoErr) {
-      OMXFFTSpec_R_F32* context =
-          static_cast<OMXFFTSpec_R_F32*>(malloc(bufSize));
-      if (omxSP_FFTInit_R_F32(context, order) != OMX_Sts_NoErr) {
-        return nullptr;
-      }
-      return context;
-    }
-    return nullptr;
-  }
-#endif
-
   void Clear() {
-#if defined(MOZ_LIBAV_FFT)
     if (mAvRDFT) {
       sRDFTFuncs.end(mAvRDFT);
       mAvRDFT = nullptr;
@@ -313,34 +197,15 @@
       sRDFTFuncs.end(mAvIRDFT);
       mAvIRDFT = nullptr;
     }
-#else
-#  ifdef BUILD_ARM_NEON
-    free(mOmxFFT);
-    free(mOmxIFFT);
-    mOmxFFT = mOmxIFFT = nullptr;
-#  endif
-    free(mKissFFT);
-    free(mKissIFFT);
-    mKissFFT = mKissIFFT = nullptr;
-#endif
   }
   void AddConstantGroupDelay(double sampleFrameDelay);
   void InterpolateFrequencyComponents(const FFTBlock& block0,
                                       const FFTBlock& block1, double interp);
-#if defined(MOZ_LIBAV_FFT)
   static FFmpegRDFTFuncs sRDFTFuncs;
   RDFTContext* mAvRDFT;
   RDFTContext* mAvIRDFT;
-#else
-  kiss_fftr_cfg mKissFFT;
-  kiss_fftr_cfg mKissIFFT;
-#  ifdef BUILD_ARM_NEON
-  OMXFFTSpec_R_F32* mOmxFFT;
-  OMXFFTSpec_R_F32* mOmxIFFT;
-#  endif
-#endif
   AlignedTArray<ComplexU> mOutputBuffer;
-  uint32_t mFFTSize;
+  uint32_t mFFTSize{};
 };
 
 }  // namespace mozilla